#Simple steps for using linux

1. Know where you are and how to get around.
2. Choose your editor.
3. Setup your project folder.
4. Get ready for the next step.

## Objectives
1. Find your home directory. Take a brief tour of your home directory. See where most programs are.
2. Create a file called .bashrc, add environemntal variable $HKSP1=your_vt_id@hokiespeed1.arc.vt.edu and $HKSP2=your_vt_id@hokiespeed2.arc.vt.edu
3. Create a working directory with the following structure.
4. Get some read files to the RawData directory


```
Project1
+-- README.md
+-- Data
|   +-- Where you save links to your raw data
|   +-- such as fastq files, genome files
+-- RawData
|   +-- Raw data for this project
+-- Figures
|   +-- Figures for this project, idealy, all figures should be generated by scripts/programs
+-- Results
|   +-- Results, summary statistics in the format of csv, tsv etc.
|   +-- For each excel file, keep one version of the raw data in raw format.
+-- Scripts
|   +-- Scripts from R, Python, Perl, Bash, Ruby
|   +-- Code from c++ or java
+-- Documents
|   +-- Daily journal, track your day to day programming/analysis progress.
|   +-- Summary for mature scripts. include input, output, working dir, command line parameters etc.
|   +-- Flow charts for your analysis routine.
Project 2 
Project 3 etc
```




## 1. Your working environment.

### 1.1 open your terminal

```
bioinfo@bioinfo-VirtualBox:~$ 

```

### 1.2 where am I now?

```
bioinfo@bioinfo-VirtualBox:~$ pwd
/home/bioinfo
```

### 1.3 what's already there for me?

```
$ ls

$ ls -lh

$ ls -a

$ df -h

$ cd /

$ ls

$ cd

$ cd /

$ cd ~/Documents

$ cd ..

$ cd -

```

### 1.4 where are the "programs" ?

```

$ ls /usr/bin

$ ls /usr/local/bin

$ which python

$ which R

```


## 2. Making your life easier by editing your .bashrc.

### 2.1 choose your editor
    * vi http://en.wikipedia.org/wiki/Vi
    * emacs http://en.wikipedia.org/wiki/Emacs
    * **nano**
    * gedit
    * notepad ++ **windows only**
    * bluefish
    * eclipse


### 2.2 open your .bashrc
```
$ cd 
$ less .bashrc
$ nano .bashrc

```

### 2.3 add environmental variables to .bashrc

go to the bottom of your .bashrc file (Ctrl+V for both win and mac) and add:

```
export HKSP1=yourid@hokiespeed1.arc.vt.edu
export HKSP2=yourid@hokiespeed2.arc.vt.edu
```
To save your editing takes three steps:

    - Ctrl + O
    - Enter
    - Ctrl + E

Check your .bashrc by _less_ or _tail_

Make sure to reload .bashrc
```
source .bashrc
```

### 2.4 login to hokiespeed

First log in:
```
ssh $HKSP1

```

# DON'T USE THE HEAD NODE FOR COMPUTATION!

After log in:
```
[songli@hslongin1 ~]$ pwd
/home/songli

$ ls -lh

$ ls -a

$ df -h

$ less .bashrc

$ ls /usr/bin

$ which python

$ which R
```

Find what's available

```
$ module list

$ module avail

$ module load R/2.14.1

$ R
R version 2.14.1 (2011-12-22)
Copyright (C) 2011 The R Foundation for Statistical Computing
ISBN 3-900051-07-0
...
...
Type 'demo()' for some demos, 'help()' for on-line help, or
'help.start()' for an HTML browser interface to help.
Type 'q()' to quit R.

> 
> q()
Save workspace image? [y/n/c]: n

[songli@hslogin1 ~]$ 

```

Leaving HokieSpeed

```
$ exit

```

## 3. Creating a good starting point.
* create a working directory for a bioinformatics project

```
$ cd
$ mkdir Research
$ cd Research
$ mkdir Project1
$ cd Project1
$ mkdir data
$ mkdir rawdata
$ mkdir scripts
$ mkdir results
$ mkdir figures
$ mkdir documents
$ echo "my first project" > README.md
```

## 4. Get ready for the next step.
### 4.1 We will download four files from this github repository for our next section:

* SampleFastq.fastq [copythislink](http://tinyurl.com/lathambio/SampleFastq.fastq)

    fastq file is the commonly used file format for next generation sequencing. 

* SampleGFF.gff [copythislink](http://tinyurl.com/lathambio/SampleGFF.gff)

    gff or gtf files are commonly used for annotate gene structures. 

* SampleSam.sam [copythislink](http://tinyurl.com/lathambio/SampleSam.sam)

    sam or bam files are used for saving (short) read alignment information. 

* TAIR10.fas [copythislink](http://tinyurl.com/lathambio/TAIR10.fas)

    fasta files are used to save DNA/RNA/Protein sequences



### 4.2 Download these files by copying the link and go to your terminal.

```
cd
cd Research
cd Project1
cd rawdata
wget http://tinyurl.com/lathambio/SampleSam.sam
wget http://tinyurl.com/lathambio/SampleGFF.gff
wget http://tinyurl.com/lathambio/SampleFastq.fastq
wget http://tinyurl.com/lathambio/TAIR10.fas

ls -lh
```




